### ISSUE TITLE
DartkCrashes happening from seeming mis-optimization on IA32 opt-count threshold

### ISSUE THREAD

**Issue by sjindel-google** [dart-lang/sdk#35454]
The following tests are "DartkCrash"-ing with seemingly related issues:

```
mirrors/deferred_mirrors_update_test
duplicate_interface_implements_test/none
prefix_unqualified_invocation_test/none
```

The stack traces look like:
```
Crash when compiling file:///b/s/w/ir/out/ReleaseIA32/generated_tests/language_2/duplicate_interface_lib.dart,
at character offset null:
NoSuchMethodError: Class 'DietListener' has no instance method 'popCharOffset'.
Receiver: Instance of 'DietListener'
Tried calling: popCharOffset()
#0      Object.noSuchMethod (dart:core/runtime/libobject_patch.dart:50:5)
#1      OutlineBuilder.endLibraryName (package:front_end/src/fasta/source/outline_builder.dart:437:5)
#2      Parser.parseLibraryName (package:front_end/src/fasta/parser/parser.dart:650:14)
#3      Parser.parseTopLevelKeywordDeclaration (package:front_end/src/fasta/parser/parser.dart:623:18)
#4      Parser.parseTopLevelDeclarationImpl (package:front_end/src/fasta/parser/parser.dart:489:14)
#5      Parser.parseUnit (package:front_end/src/fasta/parser/parser.dart:354:15)
...
```

Looking at the code, it's clear that the error is not in the FE (Dart code). The concrete receiver type should be `OutlineBuilder`, not `DietListener`.

**Comment by mraleph** [comment]
@sjindel-google please take a look at the graph and see if you can figure out what exactly is wrong with it.

**Comment by sjindel-google** [comment]
I'm working on it. I haven't managed to reproduce it, so i'm trying to get a crash and a core dump out of the buildbots.

**Comment by dcharkes** [comment]
Maybe this segfault is related: https://github.com/dart-lang/sdk/issues/35473

**Comment by aam** [comment]
Another case of this crash in `python tools/test.py -n dartk-optcounter-linux-release-ia32 language_2/prefix_unqualified_invocation_test/01`: https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket.appspot.com/8925332900502690784/+/steps/vm_tests_shard_7/0/stdout

**Comment by mkustermann** [comment]
@mraleph Could this also be caused by the the slot canonicalization problem of fields you've diagnosed?

**Comment by sjindel-google** [comment]
Most likely fixed by a62e9248 since it hasn't reproduced since January 3 (Slava's commit landed on Jan. 7).
